Remove reserved types

Baccata opened the pull request:
Remove the out-dated type-reservation list. The only things that are reserved are language keywords now. 

Closes https://github.com/disneystreaming/smithy4s/issues/1051

miguel-vila committed:
reproduce issue with name overrides

miguel-vila committed:
Merge branch 'series/0.17' into name-override-bug

miguel-vila committed:
fix reference to trait

Baccata committed:
Drop reserved types altogether.

Collision avoidance does not need for types to be reserved any more.

Baccata committed:
De-alias String to java.lang.String during Rendering

Baccata committed:
Remove more hardcoded type ascriptions in rendering

Baccata commented:
@miguel-vila, @kubukoz  although my fix works and is actually much better in the grand scheme of things (no more reserved types), it has the annoying side effect of drastically impacting the UX for AWS clients, has AWS has the annoying tendency of redefining primitive types in all its specs : 

https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/sdk-codegen/aws-models/dynamodb.json#L10134-L10136 

So I think we need to take action within this same unit of work to avoid this problem by means of a ProjectionTransformer that'd turn all simple shapes that are coming from aws-namespaces into references to the "standard" shapes (as long as the shapes we're transforming don't have any traits, and have the same names/shape-type as their targets). This ProjectionTransformer would be applied by default. 

@miguel-vila, if you want to pick this, it'd be greatly appreciated

Baccata committed:
Re-gen examples

daddykotex commented:
> @miguel-vila, @kubukoz although my fix works and is actually much better in the grand scheme of things (no more reserved types), it has the annoying side effect of drastically impacting the UX for AWS clients, has AWS has the annoying tendency of redefining primitive types in all its specs :
> 
> https://github.com/aws/aws-sdk-js-v3/blob/main/codegen/sdk-codegen/aws-models/dynamodb.json#L10134-L10136
> 
> So I think we need to take action within this same unit of work to avoid this problem by means of a ProjectionTransformer that'd turn all simple shapes that are coming from aws-namespaces into references to the "standard" shapes (as long as the shapes we're transforming don't have any traits, and have the same names/shape-type as their targets). This ProjectionTransformer would be applied by default.
> 
> @miguel-vila, if you want to pick this, it'd be greatly appreciated

Do you think they have something similar in-house? If they use those specs to generate their SDK, that'd mean, for example, that the Dynamo client works with dynamo.String rather than java.lang.String?

Baccata commented:
@daddykotex they don't have newtypes in house, so they always render the underlying type, which solves the problem (which obviously we don't want to do, because we have newtypes)

kubukoz commented:
@Baccata so the transformation would:

- replace occurrences of said types (no traits, unqualified name matches the underlying type) with the underlying type
- remove said types from the model

and it would be included as a default `smithy4sTransformations` entry? or would you make it a default in another way?

kubukoz commented on modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:
```diff
@@ -658,13 +658,13 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
+      line"override def getMessage(): $string_ = ${field.name}"
     case field if field.tpe.isResolved =>
-      line"override def getMessage(): String = ${field.name}.orNull"
+      line"override def getMessage(): $string_  = ${field.name}.orNull"
     case field if field.required =>
-      line"override def getMessage(): String = ${field.name}.value"
+      line"override def getMessage(): $string_  = ${field.name}.value"
     case field =>
-      line"override def getMessage(): String = ${field.name}.map(_.value).orNull"
+      line"override def getMessage(): $string_  = ${field.name}.map(_.value).orNull"
```
**issue:** too many spaces (causing unnecessary changes)

Baccata replied:
[a67eb61](https://github.com/disneystreaming/smithy4s/pull/1052/commits/a67eb61cc5c3567bc3646a1390fd1d17734aa8f1)

Baccata committed:
Remove un-necessary space

Baccata commented:
> replace occurrences of said types (no traits, unqualified name matches the underlying type) with the underlying type
remove said types from the model

absolutely 

> and it would be included as a default smithy4sTransformations entry? or would you make it a default in another way?

yes

kubukoz commented:
I believe @denisrosca from my team will look into this next week. @baccata given this is 0.17, I suppose you want to target **this** branch with the transformation PR, so that #1052 doesn't get released without that?

Baccata commented:
> I suppose you want to target this branch with the transformation PR, so that https://github.com/disneystreaming/smithy4s/pull/1052 doesn't get released without that? 

Indeed I do

denisrosca committed:
Flatten AWS newtypes

AWS in their smithy specs have a tendency to create newtypes for
the smithy standard shapes. As part of this change, introduce a
projection transformer that will run as a preprocessor before code
generation and will replace the newtypes (if they match some criteria)
with references to the standard smithy shapes.

For a newtype to be considered as being "replaceable" it has to match
the following criteria:
- doesn't have any traits (other than @default)
- is a simple shape
- the name matches the smithy shape (one exception is Date, as it
will be replaced with Timestamp even though it doesn't actually match).
- comes from an AWS namespace (i.e. starts with com.amazonaws)

denisrosca commented:
@kubukoz @Baccata 

I opened https://github.com/disneystreaming/smithy4s/pull/1110 to remove AWS newtypes during code generation.

denisrosca committed:
Keep traits on replaced members

denisrosca committed:
Use lightweight dynamo spec for sbt codegen test

Baccata committed:
Merge remote-tracking branch 'origin/series/0.17' into remove-reserved-types

denisrosca committed:
Merge branch 'remove-reserved-types' into flatten-aws-newtypes

denisrosca committed:
Fix formatting

denisrosca committed:
Add license headers

Baccata committed:
Merge pull request #1110 from denisrosca/flatten-aws-newtypes

Flatten AWS newtypes
